resharper/resharper-unity/src/CSharp/Feature/Services/QuickFixes/MoveQuickFixes/MonoBehaviourMoveUtil.cs

```diff
@@ -0,0 +1,263 @@
+                    var provider = DefaultUsagesProvider.Instance;
+                    var usages = provider.GetUsages(localVariableDeclaredElement,
+                        originExpression.GetContainingNode<IMethodDeclaration>().NotNull("scope != null"));
+                    originExpression.GetContainingStatement().NotNull("expression.GetContainingStatement() != null")
+                        .RemoveOrReplaceByEmptyStatement();
+                    foreach (var usage in usages)
+                    {
+                        if (usage.IsValid() && usage is IReferenceExpression node)
+                            node.ReplaceBy(factory.CreateReferenceExpression(newName));
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// If current expression is used as initializer for local variable, declared element for this variable will be returned
+        /// </summary>
+        public static IDeclaredElement GetDeclaredElementFromParentDeclaration([NotNull] ICSharpExpression expression)
+        {
+            expression.GetPsiServices().Locks.AssertReadAccessAllowed();
+            
+            var localVariableDeclaration =
+                LocalVariableDeclarationNavigator.GetByInitial(
+                    expression.GetContainingParenthesizedExpression()?.Parent as IVariableInitializer);
+            return localVariableDeclaration?.DeclaredElement;
+        }
+
+        public static string CreateBaseName([NotNull]ICSharpExpression toMove, [CanBeNull] IDeclaredElement variableDeclaration)
+        {
+            toMove.GetPsiServices().Locks.AssertReadAccessAllowed();
+            
+            var type = toMove.Type(new ResolveContext(toMove.GetPsiModule()));
+            if (type.IsUnknown)
+                type = TypeFactory.CreateTypeByCLRName("System.Object", toMove.GetPsiModule());
+            
+            string baseName =  type.GetPresentableName(CSharpLanguage.Instance);
+
+            if (variableDeclaration != null)
+            {
+                baseName = variableDeclaration.ShortName;
+            }
+            else
+            {
+                if (toMove is IInvocationExpression invocationExpression)
+                {
+                    var arguments = invocationExpression.Arguments;
+                    if (arguments.Count > 0)
+                    {
+                        var argument = arguments[0].Value;
+                        var reference = argument.GetReferences<UnityObjectTypeOrNamespaceReference>().FirstOrDefault();
+                        if (reference != null && reference.Resolve().ResolveErrorType.IsAcceptable)    
+                        {
+                            baseName = (argument.ConstantValue.Value as string).NotNull(
+                                "argument.ConstantValue.Value as string != null");
+                        }
+                    }
+                }
+            }
+
+            return baseName;
+        }
```

Comment:
It looks like this method tries to emulate the following:
```cs
INamesCollection namesCollection;
namesCollection.Add(GetNameRootFromStringArgument(), EntryOptions { Emphasis = Good ... });
namesCollection.Add(expression, EntryOptions { ... }); // this will add roots based on the expression's type and text
```

`INamesCollection` already has the capability to suggest names based on the expression's type.
It also can suggest names based on the expression's code. It's ignored here altough I'm not sure if it's intended.